Lib/_pyrepl/fancycompleter.py

```diff
@@ -0,0 +1,175 @@
+#   Copyright 2010-2025 Antonio Cuni
+#                       Daniel Hahler
+#
+#                        All Rights Reserved
+"""Colorful tab completion for Python prompt"""
+from _colorize import ANSIColors, get_colors, get_theme
+import rlcompleter
+import types
+import keyword
+
+class Completer(rlcompleter.Completer):
+    """
+    When doing someting like a.b.<tab>, display only the attributes of
+    b instead of the full a.b.attr string.
+
+    Optionally, display the various completions in different colors
+    depending on the type.
+    """
+    def __init__(
+        self,
+        namespace=None,
+        *,
+        use_colors='auto',
+        consider_getitems=True,
+    ):
+        from _pyrepl import readline
+        rlcompleter.Completer.__init__(self, namespace)
+        if use_colors == 'auto':
+            # use colors only if we can
+            use_colors = get_colors().RED != ""
+        self.use_colors = use_colors
+        self.consider_getitems = consider_getitems
+
+        if self.use_colors:
+            readline.parse_and_bind('set dont-escape-ctrl-chars on')
+        if self.consider_getitems:
+            delims = readline.get_completer_delims()
+            delims = delims.replace('[', '')
+            delims = delims.replace(']', '')
+            readline.set_completer_delims(delims)
+
+    def complete(self, text, state):
+        # if you press <tab> at the beginning of a line, insert an actual
+        # \t. Else, trigger completion.
+        if text == "":
+            return ('\t', None)[state]
+        else:
+            return rlcompleter.Completer.complete(self, text, state)
+
+    def _callable_postfix(self, val, word):
+        # disable automatic insertion of '(' for global callables
+        return word
+
+    def global_matches(self, text):
+        names = rlcompleter.Completer.global_matches(self, text)
+        prefix = commonprefix(names)
+        if prefix and prefix != text:
+            return [prefix]
+
+        names.sort()
+        values = []
+        for name in names:
+            clean_name = name.rstrip(': ')
+            if clean_name in keyword.kwlist:
+                values.append(None)
+            else:
+                try:
+                    values.append(eval(name, self.namespace))
+                except Exception as exc:
+                    values.append(None)
+        if self.use_colors and names:
+            return self.colorize_matches(names, values)
+        return names
+
+    def attr_matches(self, text):
+        expr, attr = text.rsplit('.', 1)
+        if '(' in expr or ')' in expr:  # don't call functions
+            return []
+        try:
+            thisobject = eval(expr, self.namespace)
```

Comment:
Could this be the following?
```suggestion
            thisobject = self.namespace.get(expr)
```

Reply:
no, it doesn't work in case of e.g. `a.b.c.<TAB>`

Reply:
I see. In general I like to avoid using `eval` in code (because of safety and performance reasons). Not using eval here would add a few lines of code though. Leaving the decision on this to a core dev